### PR TITLE
fix(random-sampling): content-bind storage proofs (close caller-supplied-leaf gap)

### PR DESCRIPTION
**Pull request by branarakic**
## Summary

`RandomSampling.submitProof` trusted a **caller-supplied `leaf`** that was never bound to the challenged content, and the verifier never constrained proof length — so a storage proof could verify **without the prover holding the underlying data**. This is a regression from V8 (which derived the leaf on-chain from the submitted chunk) and affects both the public and curated `_catalog` challenge branches. It's the proof-of-storage gate that authorizes reward accrual, so closing it pre-mainnet is the goal.

## Fix — content-binding + structured commitment

- **Content-binding (the fix):** `submitProof(bytes content, …)` derives `leaf = keccak256(content)` on-chain — the prover must submit the real data (public N-Triple bytes / curated `_catalog` triple bytes). A leaf that isn't backed by content can no longer pass. Both branches.
- **Structured commitment:** the public-CG root becomes `hashPair(publicRoot, privateDataHash)` — private data stays **anchored** in the on-chain root (integrity preserved) but is **never a challengeable leaf**; sampling draws only public, content-bindable leaves. The curated path content-binds the public `_catalog` (a plain tree). Single source of truth: `structuredKARootV10` in `dkg-core`.
- Prover, publisher, and chain adapter updated to the new format; `RandomSampling.json` ABI synced; the agent's legacy "no private-root anchoring" verify fallback removed.
- `RandomSampling` `_VERSION` 10.1.0 → **10.2.0**. The `_verifyV10MerkleProof` loop is unchanged (it already folds the extra `privateDataHash` sibling).

## Verification

| Suite | Result |
|---|---|
| dkg-core | 1062 ✓ (incl. the bypass-rejected crypto test) |
| random-sampling | 64 ✓ — **real on-chain e2e**: honest proof `solved=true` **and** an empty-proof/echo-root submission **reverts** against the deployed `RandomSampling.sol` |
| chain | 609 ✓ |
| publisher | structured-root consumers ✓ (v10-merkle, ack-collector, storage-ack/update, protocol-ops) |
| evm-module | RandomSampling 33 ✓ + curated 25 ✓ |
| agent | compiles ✓ |

## Cutover

Changes the on-chain commitment format + the `submitProof` ABI, so it requires an **atomic cutover** (redeploy + clear outstanding challenges + re-publish fixtures) — see `CUTOVER.md`. Pre-mainnet, no data migration.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: packages/publisher/src/merkle.ts [hunk 1]
```diff
@@ -69,27 +69,38 @@ export function computePrivateRootV10(privateQuads: Quad[]): Uint8Array | undefi
...
+/**
+ * V10 per-KA commitment root — **STRUCTURED** `hashPair(publicRoot, privateDataHash)`
+ * (replaces the former flat-mixed `publicHashes ++ privateRoots` tree). Public
+ * triples form the challengeable subtree; private data is collapsed to one
+ * sibling, bound into the root for integrity but NEVER a challengeable leaf.
+ *
+ * NOTE: name kept (`...FlatKC...`) so every recompute callsite inherits the new
+ * root with no edit; the tree is no longer flat-mixed. TODO: rename in cleanup.
+ */
 export function computeFlatKCRootV10(
```
Comment (otReviewAgent) [inline packages/publisher/src/merkle.ts:99]: The V10 publisher API now returns a structured commitment, but the exported names and aliases still say `computeFlatKCRootV10` / `computeFlatKCRoot`. The implementation comment acknowledges this as a TODO, but this function is imported broadly across publisher and agent code, so keeping the old abstraction name makes future call sites easy to misuse when reasoning about proof shape and leaf count. A clearer migration would add a `computeStructuredKCRootV10` (ideally returning root and public leaf count together) and keep the old flat name only as a deprecated wrapper during compatibility cleanup.

File: packages/random-sampling/src/proof-builder.ts [hunk 2]
```diff
@@ -19,32 +19,34 @@
...
+   * Public path: the KC's public triples. Catalog path: the `_catalog` triples.
    * Order does not matter — `V10MerkleTree` sorts + dedupes.
-   *
-   * For curated KCs (LU-11 / RFC-39) when `kind: 'ciphertext-chunks'`,
-   * pass the **raw ciphertext chunk bytes** in chunkId order instead;
-   * the builder hashes each chunk and uses the index-preserving
-   * `V10CiphertextChunksMerkleTree`.
    */
-  leaves: Uint8Array[];
+  contents: Uint8Array[];
```
Comment (otReviewAgent) [inline packages/random-sampling/src/proof-builder.ts:34]: `ProofBuilderRequest` is not modeled as a discriminated union: `contents` means public triples or catalog triples depending on `kind`, while `privateRoots` is required even though it is ignored for `kind: 'catalog'`. That creates invalid states at the type boundary and forces callers/tests to remember cross-field rules manually. Splitting this into `{ kind: 'public'; contents; privateRoots; ... } | { kind: 'catalog'; contents; ... }` would make the abstraction carry the tree-shape invariants instead of relying on comments.

**Comment by branarakic** [comment]
Addressed the review feedback and resolved the conflicts with `main`.

**Review feedback (@otReviewAgent)**
- **`ProofBuilderRequest` discriminated union** — now `{ kind: 'public'; contents; privateRoots } | { kind: 'catalog'; contents }`, so `privateRoots` only exists on the public path (no invalid states / cross-field rules to remember). Threaded through the prover (variant construction) and the worker (`BuildRequest = { taskId } & ProofBuilderRequest`; the message spreads the union).
- **Misleading `computeFlatKCRootV10` name** — added `computeStructuredKCRootV10` returning `{ root, leafCount }` atomically (so the on-chain root + count can't be paired inconsistently). `computeFlatKCRootV10` / `computeFlatKCMerkleLeafCountV10` are kept as `@deprecated` thin wrappers so the existing recompute call sites need no edit during migration.

**Conflicts** — only the `RandomSampling` `_VERSION` line collided (main → `10.1.1` for RFC-51 publishing-allocation; this PR → `10.2.0` for content-binding). RFC-51's `_calculateNodeScore` change and this PR's `submitProof` change are in different functions, so the bodies merged cleanly; resolved to `10.2.0` (supersedes; the comment preserves the RFC-51 context). ABI re-synced from the merged contract.

Post-merge + refactor: dkg-core / publisher / random-sampling build; random-sampling (65, incl. the on-chain empty-proof-revert) and EVM `RandomSampling` (33) suites green.

File: packages/publisher/src/merkle.ts (hunk 1 above)
Comment (otReviewAgent) [reply packages/publisher/src/merkle.ts:99]: The new structured helper is wrapped back into the deprecated `computeFlatKC*` API, and publish/update call sites still rely on those names and separate root/count calls. That preserves the misleading flat abstraction exactly where the change needs an atomic `{root, leafCount}` contract. Prefer migrating main callers to `computeStructuredKCRootV10` and keeping the deprecated wrappers only at external compatibility edges.

File: packages/random-sampling/test/proof-worker.test.ts [hunk 3]
```diff
@@ -32,20 +29,21 @@ describe.skipIf(!existsSync(ENTRY))('WorkerThreadProofBuilder (requires dist bui
...
-    const tree = new V10MerkleTree(leaves);
+    const contents = triples.map((t) => tripleContentV10(t.subject, t.predicate, t.object));
+    const tree = new V10MerkleTree(contents.map(keccak256));
     const expected = { merkleRoot: tree.root, merkleLeafCount: tree.leafCount };
 
     for (let chunkId = 0; chunkId < tree.leafCount; chunkId++) {
-      const inProcess = buildV10ProofMaterial(leaves, chunkId, expected);
-      const fromWorker = await builder.build({ leaves, chunkId, expected });
+      const inProcess = buildV10CatalogProofMaterial(contents, chunkId, expected);
+      const fromWorker = await builder.build({ contents, privateRoots: [], chunkId, expected, kind: 'catalog' });
```
Comment (otReviewAgent) [inline packages/random-sampling/test/proof-worker.test.ts:45]: The catalog request object includes `privateRoots`, but `ProofBuilderRequest` now makes `privateRoots` valid only for `kind: 'public'`. This object literal is not assignable to the discriminated union (`kind: 'catalog'` rejects the extra field), so TypeScript compilation of this test should fail. The same issue appears in the root-mismatch case below. Drop `privateRoots` from catalog requests.

The worker smoke test only exercises `kind: 'catalog'`. Production agent wiring uses `WorkerThreadProofBuilder` by default (`packages/agent/src/random-sampling-bind.ts:140`), while the changed public-CG path now depends on worker serialization of `contents` plus `privateRoots` and the structured sibling proof. The prover/Hardhat tests construct `RandomSamplingProver` without this worker, so a broken worker public branch would not be caught. Add a worker test for `kind: 'public'` with non-empty `privateRoots` and compare it to `buildV10ProofMaterial`/`verifyV10ProofMaterial`.

**Comment by zsculac** [comment]
Review conclusion:

I reviewed the PR with focus on the RandomSampling proof path, token leakage risk, and whether a non-admin node can cheat proof submission.

Summary:
- I did not find a working security issue in the contract change itself.
- The core change is directionally correct: `submitProof` now accepts content bytes, derives `leaf = keccak256(content)` on-chain, and verifies that leaf plus the supplied Merkle proof against the pinned challenge root.
- The old attack shape, `submitProof(challengeRootBytes, [])`, no longer works because the contract hashes the submitted bytes first. That path reverts with `MerkleRootMismatchError`.
- I did not find a direct token leakage path. Failed proofs revert before the challenge is marked solved or proof scores are incremented. `submitProof` does not directly transfer tokens.

Verification performed:
- `pnpm --dir packages/random-sampling exec vitest run --config vitest.config.ts` passed: 65 tests / 8 files, including the real Hardhat RandomSampling E2E.
- I stopped prior devnets. Default API port 9201 was occupied, so I ran a 6-node devnet on alternate ports 8555/9301/10301.
- On that devnet, a direct empty-proof attack against a live challenge using `submitProof(challengeRootBytes, [])` was rejected with `MerkleRootMismatchError`.
- The full daemon prover path on devnet did not complete because the nodes stayed in `kc-not-synced` after a confirmed publish/ACK/on-chain registration. I did not use that as proof of the honest daemon path.

Remaining issues before merge:
1. The integration seam test `packages/evm-module/test/integration/RandomSampling-submitproof-seam.test.ts` is stale and currently fails. It still expects `submitProof(merkleRoot, [])` to succeed at the old proof API. That is now correctly rejected. This should be rewritten to submit real content bytes and valid Merkle proofs.
2. Some comments/NatSpec still describe a caller-supplied leaf model. They should be updated to describe the content-bound API.
3. The mock adapter proof path is weaker than Solidity verification, so mock tests should not be treated as full proof-shape coverage.

Recommendation: the contract approach looks sound for the targeted issue, but I would update the stale integration tests and comments before merging so the repository matches the new security model.

File: packages/random-sampling/test/proof-worker.test.ts (hunk 3 above)
Comment (otReviewAgent) [reply packages/random-sampling/test/proof-worker.test.ts:45]: The worker-thread smoke now posts only `kind: 'catalog'`, so it no longer validates the changed worker protocol for the common public path (`kind: 'public'` plus `privateRoots`). A serialization or branch-selection regression in `proof-worker-entry.ts` for public structured proofs would not be caught here. Add a worker round trip for `kind: 'public'` with at least one private root and assert the proof includes the private sibling.

File: packages/chain/src/mock-adapter.ts [hunk 4]
```diff
@@ -1643,10 +1643,11 @@ export class MockChainAdapter implements ChainAdapter {
...
-    const leafHex = (typeof leaf === 'string' ? leaf : ethers.hexlify(leaf)).toLowerCase();
-    if (!/^0x[0-9a-f]{64}$/.test(leafHex)) {
-      throw new Error('Mock: submitProof leaf must be a 32-byte hex string (bytes32)');
-    }
+    // CONTENT-BINDING: derive the leaf from the submitted content, exactly as the
+    // chain does (`leaf = keccak256(content)`). The prover submits the N-Triple /
+    // catalog content bytes; an attacker can no longer echo the public root.
+    const contentBytes = typeof content === 'string' ? ethers.getBytes(content) : content;
+    const leafHex = ethers.keccak256(contentBytes).toLowerCase();
     if (expectedLeaf.toLowerCase() !== leafHex) {
```
Comment (otReviewAgent) [inline packages/chain/src/mock-adapter.ts:1651]: The mock accepts a proof after only checking `keccak256(content) === expectedLeaf`; it never folds `merkleProof` against the pinned `challenge.challengeRoot`. For structured public KCs, `submitProof(content, [])` can solve in the mock whenever the content hashes to the selected public leaf, but the real contract requires the appended private/sentinel sibling to reach `hashPair(publicRoot, privateDataHash)`. This will mask missing-sibling or wrong-proof regressions in off-chain integration tests.

File: packages/random-sampling/test/e2e-hardhat-chain.test.ts [hunk 5]
```diff
@@ -76,9 +77,8 @@ describe('Random Sampling E2E (Hardhat)', () => {
   const rawLeaves = publishQuads.map((q) =>
     hashTripleV10(q.subject, q.predicate, q.object),
   );
-  const tree = new V10MerkleTree(rawLeaves);
-  const merkleRoot = tree.root;
-  const merkleLeafCount = tree.leafCount;
+  // structured public commitment (no private payload here): hashPair(publicRoot, sentinel).
```
Comment (otReviewAgent) [inline packages/random-sampling/test/e2e-hardhat-chain.test.ts:80]: The Solidity-backed random-sampling E2E explicitly covers the no-private/sentinel case only. The new structured public proof path also has to append a real collapsed `privateDataHash` sibling, but the private-root coverage is currently limited to TS helper/publisher assertions, not a real `RandomSampling.sol` submission. Add an on-chain prover/submitProof test with `privateRoots.length > 0` so the validation evidence covers the privateQuads/private-file cutover path, not just the sentinel path.

File: packages/core/src/crypto/proof-material.ts [hunk 6]
```diff
@@ -85,49 +93,99 @@ export class V10ProofChunkOutOfRangeError extends RangeError {
...
  */
 export function buildV10ProofMaterial(
-  rawLeaves: Uint8Array[],
+  publicLeafContents: Uint8Array[],
+  privateRoots: Uint8Array[],
   chunkId: number,
   expected: V10MerkleCommitment,
 ): V10ProofMaterial {
-  const tree = new V10MerkleTree(rawLeaves);
+  const leaves = publicLeafContents.map((c) => keccak256(c));
```
Comment (otReviewAgent) [inline packages/core/src/crypto/proof-material.ts:124]: `buildV10ProofMaterial` and `buildV10CatalogProofMaterial` duplicate the same content hashing, expected root/count checks, chunk bounds checks, hash-to-content lookup, and return shaping. Since these are security-sensitive proof boundaries, duplicated mechanics increase drift risk the next time one branch changes. A small internal helper that accepts the built tree/root/leafCount and optional suffix proof siblings would keep the public/catalog differences explicit while centralizing the invariant checks.
